### DISCUSSION TITLE
How to add or remove event listener dynamically?

### DISCUSSION THREAD

**Discussion by humbleCoder1990** [solidjs/solid#1382] [Q&A]
It seems "div  onPointerDown={undefinded} "  not working.
I want to make component which get props "_onPointerDown".
If user doesn't pass props, the div should't make event listener.
Is it possible?

**Comment by atk** [comment]
As discussed on discord (https://discord.com/channels/722131463138705510/1046482922351362109/1046487807817953390), use a reactive handler that you evaluate in the event prop.

**Reply by arnfaldur** [reply]
The link does not work for me. 
It would be very useful to copy the relevant snippets or discussions here for future reference.
If it is still available.

**Reply by snnsnn** [reply]
Here is the answer from the link:

```js
onPointerDown={(ev) => handlerSignal()(ev)}
```
> When setting handlerSignal, you can only ever use the setter function style, i.e. setHandlerSignal(()=>(ev)=>...)
> And make sure it always contains a function, even if that does nothing.
> Or otherwise check in the event prop if the signal contains a function.

**Reply by trusktr** [reply]
In case people stumble here, if we translate the answer that @atk linked to, it is really this:

> Event handlers cannot be removed. Use logic inside the handler to run, or not run, different functions.

**Comment by snnsnn** [comment]
You can use element's ref and onMount hook, I used click event, you can change it to anything you like:

```typescript
import { Component, onMount } from 'solid-js';

export const Comp: Component<{ onClick: (event: MouseEvent) => void }> = (
  props,
) => {
  let divRef: HTMLDivElement;

  onMount(() => {
    if (props.onClick) {
      divRef.addEventListener('click', props.onClick);
    }
  });

  return <div ref={divRef!}>Click Here</div>;
};

export const App = () => {
  const handleClick = (event: MouseEvent) => {
    console.log(event);
  };

  return (
    <div>
      <Comp onClick={handleClick} />
    </div>
  );
};
 ```

If you don't like to use onMount, you can use div ref in its function form. This is better for adding multiple conditional handlers:

```typescript
import { Component, onMount } from 'solid-js';

export const Comp: Component<{
  onClick?: (event: MouseEvent) => void;
  onMouseOver?: (event: MouseEvent) => void;
}> = (props) => {

  function attachHandlers(el: HTMLDivElement) {
    if (props.onClick) {
      el.addEventListener('click', props.onClick);
    }

    if (props.onMouseOver) {
      el.addEventListener('mouseover', props.onMouseOver);
    }
  }

  return <div ref={attachHandlers}>Click Here</div>;
};

export const App = () => {
  const handleClick = (event: MouseEvent) => {
    console.log(event);
  };

  return (
    <div>
      <Comp onClick={handleClick} />
    </div>
  );
};
```
